language/stdlib/modules/FixedPoint32.move

```diff
@@ -44,20 +44,51 @@ module FixedPoint32 {
         assert(product <= MAX_U64, Errors::limit_exceeded(EMULTIPLICATION));
         (product as u64)
     }
+
+    /// We specify the concrete semantics of the implementation but use
+    /// an abstracted, simplified semantics for verification of callers. For the verification outcome of
+    /// callers, the actual result of this function is not relevant, as long as the abstraction behaves
+    /// homomorphic.
     spec fun multiply_u64 {
         pragma opaque;
-        include MultiplyAbortsIf;
-        ensures result == spec_multiply_u64(val, multiplier);
+        include [concrete] ConcreteMultiplyAbortsIf;
+        ensures [concrete] result == spec_concrete_multiply_u64(val, multiplier);
+        include [abstract] MultiplyAbortsIf;
+        ensures [abstract] result == spec_multiply_u64(val, multiplier);
     }
+    spec schema ConcreteMultiplyAbortsIf {
+        val: num;
+        multiplier: FixedPoint32;
+        aborts_if spec_concrete_multiply_u64(val, multiplier) > MAX_U64 with Errors::LIMIT_EXCEEDED;
+    }
+    spec define spec_concrete_multiply_u64(val: num, multiplier: FixedPoint32): num {
+        (val * multiplier.value) >> 32
+    }
+
+    /// # Abstract Semantics
+
     spec schema MultiplyAbortsIf {
         val: num;
         multiplier: FixedPoint32;
         aborts_if spec_multiply_u64(val, multiplier) > MAX_U64 with Errors::LIMIT_EXCEEDED;
     }
+
     spec define spec_multiply_u64(val: num, multiplier: FixedPoint32): num {
-        (val * multiplier.value) >> 32
+        if (multiplier.value == 0)
+            // Zero value
+            0
+        else if (multiplier.value == 1)
+            // 1.0
+            val
```

Comment:
It seems that the multiplication result does not match to the multiplier. If `multiplier.value == 1`, then the result of the multiplication is `val * (1 / 2^32)` because `multiplier` is equivalent to `(1 / 2^32)`.

`else if (multiplier.value == (2 << 32))` seems the right branching condition for the case of "val * 1".

Reply:
It does intentionally not match the multiplier because it is an abstraction and the representation is different than in the concrete case.